torch/csrc/autograd/functions/convolution.cpp

```diff
@@ -644,67 +641,64 @@ static std::unique_ptr<Tensor> compute_grad_input(
+            dilated ? params.dilation[1] : 1,
+            params.output_padding[0], params.output_padding[2], params.output_padding[1]); goto done;
     }
-  } else /* !dilated */ {
-    if (params.transposed) {
-      /* !dilated && transposed */
-      if (dim == 4) {
-        SpatialFullConvolution_updateGradInput(
+  } else {  /* Not transposed */
+    if (dim == 4) {
+      if (dilated) {
+        SpatialDilatedConvolution_updateGradInput(
             input, grad_output, grad_input.get(), weight, columns,
             kernel_size[1], kernel_size[0],
             params.stride[1], params.stride[0],
             params.padding[1], params.padding[0],
-            params.output_padding[1], params.output_padding[0]); goto done;
-      } else if (dim == 5) {
-        VolumetricFullConvolution_updateGradInput(
-            input, grad_output, grad_input.get(), weight, columns, ones,
-            params.stride[0], params.stride[2], params.stride[1],
-            params.padding[0], params.padding[2], params.padding[1],
-            params.output_padding[0], params.output_padding[2], params.output_padding[1]); goto done;
-      }
-    } else /* !transposed */ {
-      /* !dilated && !transposed */
-      if (dim == 4) {
+            params.dilation[1], params.dilation[0]); goto done;
+      } else {
+        /* CPU implementation has specialized MM kernels 
+           for non-dilated case here */
         SpatialConvolutionMM_updateGradInput(
             input, grad_output, grad_input.get(), weight, columns, ones,
             kernel_size[1], kernel_size[0],
             params.stride[1], params.stride[0],
             params.padding[1], params.padding[0]); goto done;
-      } else if (dim == 5 && input->isCuda()) {
-        VolumetricConvolution_updateGradInput(
+      }
+    } else if (dim == 5 && (input->isCuda() || dilated)) {
+        VolumetricDilatedConvolution_updateGradInput(
             input, grad_output, grad_input.get(), weight, columns,
+            kernel_size[0], kernel_size[2], kernel_size[1],
             params.stride[0], params.stride[2], params.stride[1],
-            params.padding[0], params.padding[2], params.padding[1]); goto done;
-      } else if (dim == 5) {
+            params.padding[0], params.padding[2], params.padding[1],
+            dilated ? params.dilation[0] : 1,
+            dilated ? params.dilation[2] : 1,
+            dilated ? params.dilation[1] : 1); goto done;
+    } else if (dim == 5) { /* dim == 5, CPU, non-dilated */
+        /* CPU implementation has specialized MM kernels 
+           for non-dilated case here */
         VolumetricConvolutionMM_updateGradInput(
             input, grad_output, grad_input.get(), weight, columns, ones,
             kernel_size[0], kernel_size[2], kernel_size[1],
             params.stride[0], params.stride[2], params.stride[1],
             params.padding[0], params.padding[2], params.padding[1]); goto done;
-      }
     }
   }
+
```

Comment:
Can you rearrange these conditions so that they form a full binary tree, like they used to? I did it specifically so we can be sure that we're not missing any cases (it has caused some bugs previously)